File: python/day02_cube_conundrum.py
# Day 2: Cube Conundrum
# https://adventofcode.com/2023/day/2

from typing import List
from util.input_util import read_input_file

class Game:
    def __init__(self, num:int, subgames:List[dict]):
        self.num = num
        self.subgames = subgames
    
    def hasNextSubgame(self):
        return len(self.subgames) > 0

    def getNextSubgame(self):
        return self.subgames.pop()
    
RED = "red"
BLUE = "blue"
GREEN = "green"
    
def parse_subgames(subgames:str) -> List[dict]:
    sgs = []
    for s in subgames.split("; "):
        # 3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green
        sg = {}
        for c in s.split(', '):
            # 3 blue
            n, c =  c.split(' ')
            sg[c] = int(n)
        sgs.append(sg)

    return sgs
            

def parse_lines():
    lines = read_input_file(2)
    games = []
    for l in lines:
        game, subgames = l.split(": ")
        g = Game(int(game.split()[1]), parse_subgames(subgames))
        games.append(g)

    return games
        

def isPossible(subgame, cubes):
    for c, n in subgame.items():
        if c in cubes and cubes[c] >= n:
            continue
        return False
    return True


def solution1():
    games = parse_lines()
    res = 0

    CUBES = {RED: 12, GREEN: 13, BLUE: 14}
    for g in games:
        possible = True
        while possible and g.hasNextSubgame():
            sg = g.getNextSubgame()
            possible = isPossible(sg, CUBES)
        
        if possible: res += g.num
    return res

    
def updateMaxCubes(subgame, maxCubes):
    for c, n in subgame.items():
        maxCubes[c] = max(maxCubes[c], n)

    
def solution2():
    games = parse_lines()
    res = 0

    for g in games:
        maxCubes = {RED: 0, GREEN: 0, BLUE: 0}
        while g.hasNextSubgame():
            updateMaxCubes(g.getNextSubgame(), maxCubes)

        prod = 1
        for _, n in maxCubes.items():
            prod *= n
        res += prod
    
    return res
    
    
    
if __name__ == '__main__':
    print("Part 1:", solution1())
    
    print('--------------')
    
    print("Part 2: ", solution2())

File: cpp/day02_cube_conundrum.cpp
// Day 2: Cube Conundrum
// https://adventofcode.com/2023/day/2
#include<iostream>
#include<fstream>
#include<set>
#include<cmath>
#include<deque>

using namespace std;

deque<string> splitstring(string s, string delimiter = " ") {
    deque<string> res;
    int l = delimiter.size();
    int end = s.size();
    int p = 0;
    while (p != -1 && p < end) {
        p = s.find(delimiter);
        res.push_back(s.substr(0, p));
        s.erase(0, p+l);
    }
    return res;
}

void printSubgame(unordered_map<string, int>& subgame) {
    for (auto m : subgame) {
        cout << m.first << ": " << m.second << " | ";
    }
    cout << endl;
}


class Game {
public:
    int number;
    deque<string> subgames;
    Game(int number) {
        this->number = number;
    }

    bool hasNextSubgame() {
        return !subgames.empty();
    }

    unordered_map<string, int> getNextSubgame() {
        unordered_map<string, int> game;
        string s = subgames.front();
        subgames.pop_front(); 
        // 3 blue, 1 red
        for (string c : splitstring(s, ", ")) {
            // 3 blue
            deque<string> cc = splitstring(c, " ");
            game[cc[1]] += stoi(cc[0]);
        }
        return game;
    }

    int getNumber() {
        return this->number;
    }
};

auto read_input() {
    string filename = "../inputs/day02_input.txt";
    cout << "Input file: " << filename << endl;
    string line;
    vector<Game*> games;
    ifstream input(filename);
    while (getline(input, line)) {
        int i = line.find(" ");
        int j = line.find(":");
        int num = stoi(line.substr(i, j-i));
        Game* g = new Game(num);
        g->subgames = splitstring(line.substr(j+2), "; ");
        games.push_back(g);
    }
    input.close();
    return games;
}

// Red, green or blue cubes
// secret number of cubes of each color in the bag
// goal is to figure out info about the number of cubes

// Which gamves would be possilbe with
const string RED = "red";
const string BLUE = "blue";
const string GREEN = "green";

unordered_map<string, int> BAG = {
    {RED, 12},
    {GREEN, 13},
    {BLUE, 14}
};


bool isPossible(unordered_map<string, int>& subgame, unordered_map<string, int>& bag) {
    for (auto c : subgame) {
        if (bag.count(c.first) && bag[c.first] >= subgame[c.first]) {
            continue;
        }
        return false;
    }
    return true;
}


int part1() {
    vector<Game*> games = read_input();
    int res = 0;

    for (Game* g : games) {
        bool possible = true;
        while (possible && g->hasNextSubgame()) {
            unordered_map<string, int> subgame = g->getNextSubgame();
            possible = isPossible(subgame, BAG);
        }
        if (possible) res += g->getNumber();
    }
    return res;
}


// fewest number of cubes of each color that could have made the game possible
// power of a set of cubes is equal to number of R, G, B multiplied together
// sum of power for each game

void updateMaxCubes(unordered_map<string, int>& subgame, unordered_map<string, int>& maxCubes) {
    for (auto s : subgame) {
        maxCubes[s.first] = max(maxCubes[s.first], s.second);
    }
}


int part2() {
    vector<Game*> games = read_input();
    int res = 0;
    for (Game* g : games) {
        unordered_map<string, int> maxCubes = {{RED, 0}, {BLUE, 0}, {GREEN, 0}};
        // need to track max of every color type
        while (g->hasNextSubgame()) {
            unordered_map<string, int> subgame = g->getNextSubgame();
            updateMaxCubes(subgame, maxCubes);
        }

        // Take power
        int pow = 1;
        for (auto m : maxCubes) {
            pow *= m.second;
        }
        res += pow;
    }
    return res;
}


int main() {
    auto p1 = part1();
    cout << "Part 1: " << p1 << endl;

    cout << "--------------" << endl;

    auto p2 = part2();
    cout << "Part 2: " << p2 << endl;
}